### PR TITLE
Remove auction_transaction

### PR DESCRIPTION
**Pull request by sunce86**
# Description
Part of https://github.com/cowprotocol/services/issues/2282 & https://github.com/cowprotocol/services/issues/2275

Exchanges `tx_from/tx_nonce` for `auction_id` which can be directly encoded from the settlement transaction since we shipped colocation. This allows us to remove an entire database table and a bunch of other complicated code.

I'm not yet removing the old columns in case something goes bad with this change.

# Changes

- [x] Removed `auction_transaction` table.
- [x] Redesigned `settlements` table to contain auction id

## How to test
1. db tests and e2e tests cover the new functionality.
2. Locally I've tested the db migration file like this:
- First comment out the V_58 migration file and apply old schema.
- Insert 10 settlements, of which first 5 have `auction_transaction` record (except one settlement (block_number 3), to simulate the existence of settlement from other environment staging/prod)
- Then, manually apply migration script V_58
- Observed the `settlements` table,`auction_id` were populated as expected.

### CODE REVIEW

**Review by fleupold — COMMENTED** [review]
I like the change of the table turning Option<from/nonce> to an Option<AuctionID>.

However, I think we should explore populating the auction id field already in the event handler that inserts the original settlement (it should have access to all required information). This way we don't need to create the extra "auction_kind" column which I find quite confusing (as the db values don't match the values on the struct) and don't run into the issue of first requiring one event handler to run before the other one does.

I'm not sure what this means for transitioning, but we could for a while have both systems run in parallel (and assert that the upsert doesn't really change any values other than at the moment of the release when it may be catching up with old auctions).

**Comment by sunce86** [comment]
> However, I think we should explore populating the auction id field already in the event handler that inserts the original settlement (it should have access to all required information).

I did a detailed exploration of this before starting to work on this issue and it can't be done without extra web3 call to get the transaction and decode the `auction_id`. The best solution would be to add the `auction_id` to the settlement event data when the new version of GPv2Settlement contract is deployed.

**Comment by fleupold** [comment]
> I did a detailed exploration of this before starting to work on this issue and it can't be done without extra web3 call to get the transaction and decode the auction_id

We are doing this in `OnSettlementEventUpdater` updater anyways and event processing is already `async` so I don't think it changes much honestly. In fact we could probably call `OnSettlementEventUpdater.update(tx_hash).await` with minimal code changes from the existing event processing code today (although the dependency injection would be quite weird).

My suggested plan of action would be:
1. Change nonce/from for auction ID but keep the current delay (this get rid of a lot of code and is a strict improvement)
2. Create a new domain component that has the required dependencies (persistence, Ethereum) and implements `EventStoring` for all events that we care about. Event processing should be domain logic imo.
3. Move the current "OnSettlementEventUpdater" logic in this component also
4. Remove 64 block delay and instead invoke update logic when the settlement event is being processed.

Wdyt?

**Comment by fleupold** [comment]
I removed the `auction_kind` enum and involved migration logic (meaning we still fetch 64 block old auctions without an auction id and amend their information). I plan to tackle those in another PR (first moving event handling into a domain component and then combining the two handlers).

Could I get a review please.

File: crates/database/src/settlements.rs
Comment (sunce86) [inline crates/database/src/settlements.rs:68]: this now means `already_processed` instead of `data_exists`, I guess

**Review by sunce86 — COMMENTED** [review]
Once merged, I think the `OnSettlementEventUpdater` will try to process the first `settlement` without `auction_id` which will be some pre-colocation settlement for which the recovery of `auction_id` from calldata will fail and it will be stuck.

**Comment by fleupold** [comment]
> Once merged, I think the OnSettlementEventUpdater will try to process the first settlement without auction_id which will be some pre-colocation settlement for which the recovery of auction_id from calldata will fail and it will 

You are right! Pre-colocation production transactions would fail on staging and vice versa. Do you think hardcoding a min block (past colocation launch) would be an ok workaround for this?

**Comment by sunce86** [comment]
> this

I think that's ok and we can remove it at a later stage of refactoring.

**Comment by fleupold** [comment]
Would there be an issue with populating a fake auction ID (e.g. -1) for items that are far in the past and cannot be matched?

**Comment by sunce86** [comment]
> Would there be an issue with populating a fake auction ID (e.g. -1) for items that are far in the past and cannot be matched?

I think theoretically this could also happen when solvers start running their own driver (until we have a new smart contract version that enforces the existence of auction_id) so we definitely want to overcome this error and continue processing other settlements. 

From the top of my head, I don't think -1 would cause something bad (would have to go through the codebase and double check all `JOIN`s though)

**Review by sunce86 — COMMENTED** [review]
Approving with comment (since I cant approve my own PR 😄 )

The only thing I noticed is that we changed the way we handle `InvalidSelector` error on decoding the transaction, but I think this can't happen after colocation (it could happen before colocation because we saved `auction_transaction` entry during competition time).

**Review by fleupold — APPROVED** [review]

**Pull request merged**
